### PR TITLE
[Core] Adding ReadMaterialProcess in cpp

### PR DESCRIPTION
**Pull request by loumalouomega**
@marandra did most of the work.

### CODE REVIEW

**Comment by loumalouomega** [comment]
I forgot to change the name of the tests on test_KratosCore.py, @marandra can you please do it?

File: kratos/tests/materials.json
Comment (philbucher) [inline kratos/tests/materials.json]: Why do you remove this?
I think this is an important check

Or does the utility not support this?

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: There is no error checking if the file exists, could be read ...

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: :+1:

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: leftover debug output

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: I would pass by reference to avoid the local copies

File: kratos/utilities/read_materials_utility.hpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.hpp]: header files are usually *.h

File: kratos/utilities/read_materials_utility.hpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.hpp]: I think some of the includes are not necessary, same in the cpp

**Comment by loumalouomega** [comment]
I just checked, is OK, sorry for the noise

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: This should not be necessary, if it is then there is sth wrong with the Parameters

I implemented this function and I am quite sure you should not do this

This can silently change the variable type!

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: same as for Matrix

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: I think you should throw an error if the size does not match

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: same as for the 6d version

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: I think these transformations are very dangerous!

silently changing is really not a good idea IMO

the user should specify the correct type

**Review by philbucher — DISMISSED** [review]
Overall I am really in favor of this PR but there are some rather serious issues IMO, see comments

**Comment by philbucher** [comment]
btw I guess you want to use it in StructuralMechanics, this will be a separate PR?

File: kratos/tests/materials.json
Comment (loumalouomega) [reply kratos/tests/materials.json]: Huuuum, @marandra after your string clean up does it work?

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: True

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Cleaning

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Really?, I wanted to be sure

**Comment by loumalouomega** [comment]
Better when this is done

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Well, this was a @RiccardoRossi  suggestion

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Well, that can be disussed

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Idem

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: I think it does not the conversion

https://github.com/KratosMultiphysics/Kratos/blob/c8cde04ff759696d3711ff69e7349f87e10ba4b4/kratos/includes/kratos_parameters.h#L366

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: I added a check

File: kratos/tests/materials.json
Comment (marandra) [reply kratos/tests/materials.json]: This change wasn't intended for the PR, I was just testing @loumalouomega change to allow for missing constitutive law. However, I do think we should update the testfile to test it.

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [reply kratos/utilities/read_materials_utility.cpp]: I will not block bcs of this, but I really don't think we should give users opportunities to become sloppy with input
Who knows which problems this can create down the road...

Your call, just keep it in mind

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [reply kratos/utilities/read_materials_utility.cpp]: Have a look at the tests I wrote:
https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_kratos_parameters.py#L473-L582

I struggled a lot with this so I am sure it is working as intended
If sth is a Matrix then IsVector will return false and vise-versa

Also if the user defines a matrix then you are doing some sort of conversion, seems hacky to me

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [reply kratos/utilities/read_materials_utility.cpp]: see comment above

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [reply kratos/utilities/read_materials_utility.cpp]: I think if the size != 6 it is undefined behavior that should not be allowed

we could even think abt implementing IsArray3d / IsArray6d in the Parameters

Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: I agree, but @RiccardoRossi tell me to do conversions in order to give more freedom to the user. I agree to define arrays on the Parameters

Comment (philbucher) [reply kratos/utilities/read_materials_utility.cpp]: ok then lets wait for @RiccardoRossi

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: Btw [this check](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/read_materials_process.py#L154-L157) is missing, can you please add it?
we had some problems with this in the past
Edit: corrected link

File: kratos/tests/materials.json
Comment (marandra) [reply kratos/tests/materials.json]: just reverted it to the original version

File: kratos/tests/test_materials_input.py
Comment (philbucher) [inline kratos/tests/test_materials_input.py:7]: why did you remove the StructuralMechanics here?

It is still depending on it (bcs of the const_law being used)

Comment (loumalouomega) [reply kratos/tests/test_materials_input.py:7]: Ups, true

**Comment by RiccardoRossi** [comment]
Guys, I need some time to review this...

**Comment by loumalouomega** [comment]
Solved

File: kratos/utilities/read_materials_utility.cpp
Comment (marandra) [inline kratos/utilities/read_materials_utility.cpp]: Here it should be .end() - .begin(). Anyways, we already tried this method to test for previous Variables and Tables and it doesn't work. I couldn't find a simple way to test it.

**Comment by marandra** [comment]
@loumalouomega I can't compile the code, I am getting the following error:
```
read_materials_utility.cpp:99:101: error: invalid operands to binary expression ('iterator' (aka 'Kratos::Parameters::iterator_adaptor') and 'iterator')
    KRATOS_INFO_IF("::[Reading materials process]:: Property", (data["Material"]["Variables"].end() - data["Material"]["Variables"].begin()) > 0 && p_prop->HasVariables()) << std::to_string(property_id) << "already has variables." << std::endl;
```
Last time we tried using .end() - .begin(), it didn't, so unless something was modified in the implementation of the iterator, I don't expect this test to work

**Comment by loumalouomega** [comment]
@marandra remember that I cannot approve this because I created the PR

**Comment by marandra** [comment]
I know, I shouldn't approve either. I'm counting on @philbucher and @RiccardoRossi

**Comment by philbucher** [comment]
I removed my blocker, approval I leave to @RiccardoRossi 

I raised my concerns, @RiccardoRossi please consider them when reviewing

File: kratos/utilities/read_materials_utility.h
Comment (philbucher) [inline kratos/utilities/read_materials_utility.h]: I think there are still unnecessary includes

**Comment by marandra** [comment]
I finished with my pending changes. Ready for review from my side.
Edit: Thanks @philbucher for your comments.

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: Params

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: Materials

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: rLine

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: Data

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: r_model_part

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: r_elements_array

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: r_conditions_array

File: kratos/utilities/read_materials_utility.h
Comment (philbucher) [inline kratos/utilities/read_materials_utility.h]: Data

File: kratos/utilities/read_materials_utility.h
Comment (philbucher) [inline kratos/utilities/read_materials_utility.h]: Materials

File: kratos/utilities/read_materials_utility.h
Comment (philbucher) [inline kratos/utilities/read_materials_utility.h]: rLine

**Review by philbucher — COMMENTED** [review]
Some comments on variable naming

**Comment by loumalouomega** [comment]
Ping @KratosMultiphysics/technical-committee

File: kratos/utilities/read_materials_utility.cpp
Comment (pooyan-dadvand) [reply kratos/utilities/read_materials_utility.cpp]: I understand that this is for read` young_modulus = 1000000` without getting an error. Isn't it?

Comment (pooyan-dadvand) [reply kratos/utilities/read_materials_utility.cpp]: You are repeating these ifs in many places. It would be better to add a method ExtractDouble and move everything there.

File: kratos/utilities/read_materials_utility.cpp
Comment (pooyan-dadvand) [reply kratos/utilities/read_materials_utility.cpp]: +1 for throwing an error.

File: kratos/utilities/read_materials_utility.cpp
Comment (RiccardoRossi) [reply kratos/utilities/read_materials_utility.cpp]: if you do 
     
       param.GetDouble()

and it is a "number" than it should convert it transparently, doing the casting internally.

@philbucher i think we should accept that if someone does not put the . we can still convert integer to double...

File: kratos/utilities/read_materials_utility.cpp
Comment (RiccardoRossi) [reply kratos/utilities/read_materials_utility.cpp]: +1 for the error. I could agree about having IsArray3, as it is a common usecase, but i would definitely not want IsArray6 ...

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Where is this conversion done?

https://github.com/KratosMultiphysics/Kratos/blob/8218d2689a6d2a55bb52b33c8ca7a80091d62d34/kratos/includes/kratos_parameters.h#L330

**Comment by loumalouomega** [comment]
What is missing in order to merge this?

**Comment by philbucher** [comment]
- errors for array3d/6d I think, see review comments
- removing isvector/ismatrix from the other one => as I wrote this is checked internally
- maybe simplify the many ifs according to pooyans comment

Plus I would add in the tests if the conversions you do (e.g. bool to int etc) are working

**Comment by loumalouomega** [comment]
Ok, I will take a look later

**Comment by loumalouomega** [comment]
Done

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp:173]: Still no errors for wrong sizes

Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp:173]: Ok, I understood that no checks at all

Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp:173]: One sec

Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp:173]: Done

**Comment by loumalouomega** [comment]
Done

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: shouldn't this be `++iter`?

File: kratos/utilities/read_materials_utility.cpp
Comment (philbucher) [inline kratos/utilities/read_materials_utility.cpp]: same

**Review by philbucher — DISMISSED** [review]
From my side it is ok now, I am approving since none else does

Please wait until tmr with merging in case @RiccardoRossi or @pooyan-dadvand still have sth to say

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp]: Huuuuum

**Review by philbucher — DISMISSED** [review]

File: kratos/utilities/read_materials_utility.cpp
Comment (pooyan-dadvand) [inline kratos/utilities/read_materials_utility.cpp:124]: Why you set properties by the model_part? This seems to be a important restriction respect to the current use of Properties per element and conditions.

I was expecting that by reading the material you would update the properties without affecting to whom they are assigned.

File: kratos/utilities/read_materials_utility.cpp
Comment (pooyan-dadvand) [inline kratos/utilities/read_materials_utility.cpp:130]: Same for conditions.

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I have blocked it just to reconfirm the intention of the Read material. In my opinion we should have the possibility to only update the properties.

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [reply kratos/utilities/read_materials_utility.cpp:124]: This process is mimicking what we are currently doing with the read_materials_process.py. What do you suggest?, taking into account that the elements and conditions may not have assigned any property at all in that moment (this is done just after reading the model part)

**Comment by pooyan-dadvand** [comment]
I would just create/update the properties assuming that they are assigned correctly to the elements and conditions (for example in mdpa)

@RiccardoRossi any comment?

**Comment by philbucher** [comment]
@pooyan-dadvand he is creating the properties if they don't exist 

this is used in StructuralMechanics very extensively in order to be able to change the properties through the `StructuralMaterials.json` instead of having to modify the mdpa
The elements and conditions get (empty) dummy properties that are exchanged in this utility

here is an example of `StructuralMaterials.json`
~~~json
{
    "properties" : [{
        "model_part_name" : "Parts_Parts_Auto1",
        "properties_id"   : 1,
        "Material"        : {
            "constitutive_law" : {
                "name" : "LinearElasticPlaneStress2DLaw"
            },
            "Variables"        : {
                "THICKNESS"     : 1.0,
                "DENSITY"       : 7850.0,
                "YOUNG_MODULUS" : 206900000000.0,
                "POISSON_RATIO" : 0.29
            },
            "Tables"           : {}
        }
    }]
}

~~~

File: kratos/utilities/read_materials_utility.cpp
Comment (RiccardoRossi) [reply kratos/utilities/read_materials_utility.cpp:124]: @pooyan, this is not so. What we did and has been in place since more or less last summer is that this assigns a given property to a submodelpart, changing the property that is assigned to that.

It is done so that the assignement is completely governed by the submodelparts (that is, in line with having mdpa to only contain connectivity information and groups)

File: kratos/utilities/read_materials_utility.cpp
Comment (RiccardoRossi) [reply kratos/utilities/read_materials_utility.cpp:130]: same comment

**Review by RiccardoRossi — COMMENTED** [review]
to me this looks ok. I am approving, pending for pooyan to release the blocking

**Comment by loumalouomega** [comment]
What do you suggest to to do?, the properties are dummy, then if I change the property of a submodelpart this will not affect the elements and conditions

**Comment by loumalouomega** [comment]
Done @RiccardoRossi

File: kratos/tests/test_materials_input.py
Comment (RiccardoRossi) [reply kratos/tests/test_materials_input.py:7]: the question is how the test could run without it...

**Review by RiccardoRossi — COMMENTED** [review]
@pooyan-dadvand can we get this through?

**Comment by loumalouomega** [comment]
Not reading fluid variables

**Comment by RiccardoRossi** [comment]
@loumalouomega but...was it giving an error?

**Comment by loumalouomega** [comment]
No, but if the Fluid is not compiled it skips the test

**Comment by marandra** [comment]
I am not convinced about the test depending on two applications, skipping it if they are not available. In my case I had to compile a complete application just to pass this test.

I would simplify the test to only use core variables. 

edit: grammar

File: kratos/utilities/read_materials_utility.cpp
Comment (pooyan-dadvand) [reply kratos/utilities/read_materials_utility.cpp:124]: @RiccardoRossi you are mentioning another pooyan which is not me! we will talk about it in @KratosMultiphysics/technical-committee

**Comment by RiccardoRossi** [comment]
@pooyan-dadvand this PR only mimics what is already done in cpp, as such i would approve it "as it is" (since i think it is needed for example by @marandra )

if you want to propose a change in the behaviour, we can definitely have that discussion but it belongs to a separate PR...  in short, i don't see reasons for keeping this blocked

**Comment by RiccardoRossi** [comment]
I should have said "mimics what is done in PYTHON"

**Review by pooyan-dadvand — APPROVED** [review]

**Comment by pooyan-dadvand** [comment]
I understand that you have migrated the python version but I'm still not convinced by the way it works and I think we should rethink this one. I will add a reminder issue.

**Comment by loumalouomega** [comment]
Thanks

File: kratos/utilities/read_materials_utility.cpp
Comment (loumalouomega) [inline kratos/utilities/read_materials_utility.cpp:139]: @josep-m-carbonell This only changed one line, not need to define the Create method

**Pull request merged**
